noir-projects/noir-protocol-circuits/crates/types/src/merkle_tree/indexed_tree.nr

```diff
@@ -112,14 +113,14 @@ pub fn insert<Value, Leaf, let TreeHeight: u32>(
     build_insertion_leaf: fn(Value, Leaf) -> Leaf,
 ) -> AppendOnlyTreeSnapshot
 where
-    Value: Eq + Empty,
-    Leaf: Hash + Empty,
+    Value: IndexedTreeLeafValue,
+    Leaf: IndexedTreeLeafPreimage,
 {
```

Comment:
I'm going to refactor in a followup the indexed tree insertions code using the leaf and preimage traits. the original code was done when traits weren't a thing so we were passing callbacks to the functions, which is uglier!